src/main/scala/wordbots/Semantics.scala

```diff
@@ -0,0 +1,293 @@
+package wordbots
+
+import scala.collection.mutable
+
+// scalastyle:off number.of.methods
+object Semantics {
+  sealed trait ParseNode extends Product
+  sealed trait AstNode extends ParseNode {
+    def depthFirstTraverse: Seq[AstNode] = {
+      val frontier = mutable.Queue(this)
+      val nodesTraversed = mutable.ListBuffer[AstNode]()
+
+      while (frontier.nonEmpty) {
```

Comment:
Why iterative instead of recursive for this DFS traversal?

Reply:
I tried writing it recursively but it was kind of a mess. Wanna try?